File: src/worker-parser-rust/src/azure/credential.rs
//! Managed identity credential provider.
//!
//! Acquires OAuth2 access tokens from the Azure Instance Metadata Service (IMDS)
//! or the Container App managed identity endpoint, caching them until near-expiry.

use std::sync::Arc;
use std::time::{Duration, Instant};

use reqwest::Client;
use serde::Deserialize;
use tokio::sync::Mutex;

/// Default resource scope for Azure management / data-plane operations.
const DEFAULT_RESOURCE: &str = "https://management.azure.com/";

/// How many seconds before actual expiry we consider the token stale.
const EXPIRY_BUFFER_SECS: u64 = 120;

/// An access token with its expiration time.
#[derive(Debug, Clone)]
pub struct AccessToken {
    pub token: String,
    pub expires_at: Instant,
}

/// Acquires tokens via Azure Managed Identity (IMDS or Container Apps).
#[derive(Clone)]
pub struct ManagedIdentityCredential {
    client: Client,
    client_id: Option<String>,
    cache: Arc<Mutex<Option<AccessToken>>>,
}

#[derive(Deserialize)]
struct ImdsTokenResponse {
    access_token: String,
    expires_in: String,
}

impl ManagedIdentityCredential {
    /// Create a new credential.  If `client_id` is non-empty it will be passed
    /// as the `client_id` query parameter (user-assigned managed identity).
    pub fn new(client_id: Option<String>) -> Self {
        Self {
            client: Client::new(),
            client_id: client_id.filter(|s| !s.is_empty()),
            cache: Arc::new(Mutex::new(None)),
        }
    }

    /// Get a valid access token for the given `resource`, refreshing if needed.
    pub async fn get_token(&self, resource: &str) -> Result<String, CredentialError> {
        {
            let guard = self.cache.lock().await;
            if let Some(ref tok) = *guard
                && Instant::now() < tok.expires_at
            {
                return Ok(tok.token.clone());
            }
        }
        let tok = self.acquire_token(resource).await?;
        let token_str = tok.token.clone();
        {
            let mut guard = self.cache.lock().await;
            *guard = Some(tok);
        }
        Ok(token_str)
    }

    /// Get a valid access token for the default management resource.
    pub async fn get_default_token(&self) -> Result<String, CredentialError> {
        self.get_token(DEFAULT_RESOURCE).await
    }

    async fn acquire_token(&self, resource: &str) -> Result<AccessToken, CredentialError> {
        // Try Container Apps managed identity endpoint first, fall back to IMDS.
        let identity_endpoint = std::env::var("IDENTITY_ENDPOINT").ok();
        let identity_header = std::env::var("IDENTITY_HEADER").ok();

        let resp = if let (Some(endpoint), Some(header)) = (identity_endpoint, identity_header) {
            // Container Apps / App Service managed identity
            let mut url = format!("{endpoint}?api-version=2019-08-01&resource={resource}");
            if let Some(ref cid) = self.client_id {
                url.push_str(&format!("&client_id={cid}"));
            }
            self.client
                .get(&url)
                .header("X-IDENTITY-HEADER", header)
                .send()
                .await
                .map_err(|e| CredentialError::Request(e.to_string()))?
        } else {
            // IMDS endpoint (VM, VMSS, etc.)
            let mut url = format!(
                "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource={resource}"
            );
            if let Some(ref cid) = self.client_id {
                url.push_str(&format!("&client_id={cid}"));
            }
            self.client
                .get(&url)
                .header("Metadata", "true")
                .send()
                .await
                .map_err(|e| CredentialError::Request(e.to_string()))?
        };

        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(CredentialError::TokenAcquisition(format!(
                "HTTP {status}: {body}"
            )));
        }

        let body: ImdsTokenResponse = resp
            .json()
            .await
            .map_err(|e| CredentialError::TokenAcquisition(e.to_string()))?;

        let expires_in: u64 = body.expires_in.parse().unwrap_or(3600);
        let expires_at =
            Instant::now() + Duration::from_secs(expires_in.saturating_sub(EXPIRY_BUFFER_SECS));

        Ok(AccessToken {
            token: body.access_token,
            expires_at,
        })
    }
}

/// Credential errors.
#[derive(Debug, thiserror::Error)]
pub enum CredentialError {
    #[error("HTTP request failed: {0}")]
    Request(String),
    #[error("token acquisition failed: {0}")]
    TokenAcquisition(String),
}

File: src/worker-graph-builder-go/internal/azure/credential/credential.go
// Package credential provides Azure Managed Identity token acquisition and caching.
package credential

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"sync"
	"time"
)

const (
	expiryBufferSecs   = 120
	tokenRequestTimeout = 30 * time.Second
)

// AccessToken holds a bearer token and its expiry time.
type AccessToken struct {
	Token     string
	ExpiresAt time.Time
}

// ManagedIdentityCredential acquires OAuth2 tokens from Azure Managed Identity.
// It supports Container Apps (IDENTITY_ENDPOINT + IDENTITY_HEADER env vars)
// and falls back to the standard IMDS endpoint.
// Tokens are cached per-resource with a 120-second early-expiry buffer.
type ManagedIdentityCredential struct {
	client   *http.Client
	clientID string
	mu       sync.Mutex
	cache    map[string]AccessToken
}

type imdsTokenResponse struct {
	AccessToken string `json:"access_token"`
	ExpiresIn   string `json:"expires_in"`
}

// New creates a ManagedIdentityCredential. Pass a non-empty clientID for
// user-assigned managed identity; pass "" for system-assigned.
func New(clientID string) *ManagedIdentityCredential {
	return &ManagedIdentityCredential{
		client: &http.Client{
			Timeout: tokenRequestTimeout,
		},
		clientID: clientID,
		cache:    make(map[string]AccessToken),
	}
}

// GetToken returns a valid access token for the given resource scope,
// refreshing from the identity endpoint when the cached token is near-expiry.
func (c *ManagedIdentityCredential) GetToken(resource string) (string, error) {
	c.mu.Lock()
	if tok, ok := c.cache[resource]; ok && time.Now().Before(tok.ExpiresAt) {
		c.mu.Unlock()
		return tok.Token, nil
	}
	c.mu.Unlock()

	tok, err := c.acquireToken(resource)
	if err != nil {
		return "", err
	}

	c.mu.Lock()
	c.cache[resource] = tok
	c.mu.Unlock()

	return tok.Token, nil
}

func (c *ManagedIdentityCredential) acquireToken(resource string) (AccessToken, error) {
	identityEndpoint := os.Getenv("IDENTITY_ENDPOINT")
	identityHeader := os.Getenv("IDENTITY_HEADER")

	var req *http.Request
	var err error

	if identityEndpoint != "" && identityHeader != "" {
		// Container Apps / App Service managed identity
		rawURL := fmt.Sprintf("%s?api-version=2019-08-01&resource=%s", identityEndpoint, resource)
		if c.clientID != "" {
			rawURL += "&client_id=" + c.clientID
		}
		req, err = http.NewRequest(http.MethodGet, rawURL, nil)
		if err != nil {
			return AccessToken{}, fmt.Errorf("build container app token request: %w", err)
		}
		req.Header.Set("X-IDENTITY-HEADER", identityHeader)
	} else {
		// Standard IMDS endpoint
		rawURL := fmt.Sprintf(
			"http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=%s",
			resource,
		)
		if c.clientID != "" {
			rawURL += "&client_id=" + c.clientID
		}
		req, err = http.NewRequest(http.MethodGet, rawURL, nil)
		if err != nil {
			return AccessToken{}, fmt.Errorf("build IMDS token request: %w", err)
		}
		req.Header.Set("Metadata", "true")
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return AccessToken{}, fmt.Errorf("token request failed: %w", err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return AccessToken{}, fmt.Errorf("token acquisition failed HTTP %d: %s", resp.StatusCode, body)
	}

	var parsed imdsTokenResponse
	if err := json.Unmarshal(body, &parsed); err != nil {
		return AccessToken{}, fmt.Errorf("parse token response: %w", err)
	}

	var expiresIn int64
	if _, scanErr := fmt.Sscanf(parsed.ExpiresIn, "%d", &expiresIn); scanErr != nil || expiresIn <= 0 {
		expiresIn = 3600
	}

	effectiveSecs := expiresIn - expiryBufferSecs
	if effectiveSecs < 0 {
		effectiveSecs = 0
	}
	expiresAt := time.Now().Add(time.Duration(effectiveSecs) * time.Second)

	return AccessToken{
		Token:     parsed.AccessToken,
		ExpiresAt: expiresAt,
	}, nil
}
